Add spell check capability

playpeace opened the pull request:
- Project specific words can be added to a dictionary
- Any typos are treated as build errors

Signed-off-by: Sumit Ahuja <ext-sahuja@luxoft.com>

tobsan commented on docs/check_spelling.sh:
```diff
@@ -0,0 +1,185 @@
+# a non-English language
+lang=en
+
+project_dict="./.aspell.en.pws"
```
Is there a specific reason why the dictionary is a dot-file? I'm thinking for usability reasons it would make sense to just call it `aspell.en.pws` (without the leading dot) or something like that.

playpeace replied:
Done.

tobsan commented on docs/check_spelling.sh:
```diff
@@ -0,0 +1,185 @@
+project_dict="./.aspell.en.pws"
+
+# Extension for files to be spell checked
+file_extension=rst
```
If it is possible to use several extensions, there are some index files that are called `.rsti` in the blueprint.

playpeace replied:
Now supports multiple extensions

tobsan commented on docs/check_spelling.sh:
```diff
@@ -0,0 +1,185 @@
+# Use aspell with no backup mode
+aspell_cmd="aspell --dont-backup --personal=$project_dict --lang=$lang"
+
+custom_dicts=""
```
What does this mean? A comment wouldn't hurt!

playpeace replied:
Done

tobsan commented on docs/check_spelling.sh:
```diff
@@ -0,0 +1,185 @@
+
+custom_dicts=""
+
+misspelled_wordlist=()
```
Is this for autocorrect? A comment wouldn't hurt here either.

playpeace replied:
Yes it is. Updated comments.

tobsan commented on docs/check_spelling.sh:
```diff
@@ -0,0 +1,185 @@
+shift $(($OPTIND - 1))
+
+
+add_additional_dicts
```
Some short comments on why things are run in a certain order would be good here.

playpeace replied:
Added.

playpeace commented:
@tobsan @rpannek I can build successfully on my local, but it fails here.
Do you have any idea why ?

tobsan commented on Jenkinsfile:
```diff
@@ -26,7 +26,7 @@ pipeline {
             steps {
                 script {
-                    String buildParams = "-DENABLE_PDF=OFF "
+                    String buildParams = "-DENABLE_PDF=OFF -DPERFORM_SPELL_CHECK=ON "
```
Even though spell check defaults to on, I like the verbosity here!

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+aab
+abb
+addr
+aditional
```
misspelled word

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+cp
+cpp
+cpu
+crosscompiling
```
should be cross-compiling

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+CXX
+dac
+DCMAKE
+dee
```
Misspelled word?

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+DL
+DockerCI
+Dockerfile
+dockerinstall
```
Shouldn't it be space here?

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+GenericLinuxOsType
+gerrit
+Gerrit
+gitclone
```
should be space?

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+preinstalled
+PREMIRRORONLY
+prepopulate
+prerequesits
```
misspelled word

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+PREMIRRORONLY
+prepopulate
+prerequesits
+presant
```
Misspelled word

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+prepopulate
+prerequesits
+presant
+PRIOTIRY
```
Misspelled word

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+Repo
+repos
+repotool
+reproducability
```
One of these are misspelled

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+URI
+url
+Url
+usefull
```
Misspelled word

tobsan commented on docs/aspell.en.pws:
```diff
@@ -0,0 +1,313 @@
+vga
+VirtualBox
+virtualization
+virtulized
```
Misspelled word

playpeace replied:
I was planning to do fix them in another commit :)

tobsan reviewed (approved):
Looks good. I commented on some words that are misspelled. Do we want to fix them in this commit or do that later?

playpeace commented:
I was planning to fix the misspelling in another commit :)

playpeace commented:
@tobsan Could you merge this if it looks okay? I suppose I won't have write access to this.

jonte commented on README.md:
```diff
@@ -60,6 +61,35 @@ which contain a toctree need to have the suffix `.trst` instead of just `.rst`.
+to verify the spellings and causes the build to fail in case of any typos.
+
+The project, which uses this blueprint, should have its own custom dictionary,
+similar to the one in blueprint (aspell.en.pws). A link to the check_spelling.sh
```
I'm guessing that the location of the project specific dictionary is important in order for the `check_spelling.sh` script to find it? Can you mention where to put the dict for project specific checks?

jonte replied:
Also, I believe the location of the symlink is also important (providing that you try finding the project specific dict based on the location of the symlink) - can you mention where the symlink should be created in the project specific case?

jonte commented on README.md:
```diff
@@ -60,6 +61,35 @@ which contain a toctree need to have the suffix `.trst` instead of just `.rst`.
+
+The spell checker has various modes and options, which can be seen by:
+
+    ./check_spelling.sh -usage
```
Can we have double dashes for `--usage` ? Also, the comvention is to call such a flag `--help` - but I am OK w/ usage.

If you want a single dash altenrative, I suggest `-u` or `-h`.

playpeace replied:
Guess we should support both short and long options.

jonte commented on README.md:
```diff
@@ -60,6 +61,35 @@ which contain a toctree need to have the suffix `.trst` instead of just `.rst`.
+
+    ./check_spelling.sh -a
+
+The dictionary can also be modified manually, by adding words to the end of the
```
Please add a note on increasing the word count (top of the aspell dict file) when adding new words manually as well. I am not sure what the significance is of the word count, but aspell seems to maintain that number when aspell itself adds words.

playpeace replied:
This is not really needed; The word count is more like a hint.

http://aspell.net/man-html/Format-of-the-Personal-and-Replacement-Dictionaries.html#Format-of-the-Personal-and-Replacement-Dictionaries

jonte replied:
Ah! Good info. Thanks :)

jonte commented on docs/CMakeLists.txt:
```diff
@@ -47,6 +47,16 @@ configure_file(
     @ONLY)
 
+add_custom_target(spell-check
+    COMMAND /bin/bash "${CMAKE_CURRENT_SOURCE_DIR}/check_spelling.sh" "-q" "${CMAKE_CURRENT_SOURCE_DIR}"
```
Maybe we can make the `check_spelling.sh` script executable, and add a hashbang at the top of it, then we don't need to invoke "/bin/bash" manually here.

playpeace replied:
Sure, there's already a hashbang, will update the cmake file.

rpannek replied:
Also not all systems have bash installed in `/bin/bash` and I'd advise to use `/usr/bin/env bash` instead, especially in the shebang line.

jonte commented on docs/check_spelling.sh:
```diff
@@ -0,0 +1,231 @@
+#!/bin/bash
```
I would add some copyright here.

jonte commented on docs/check_spelling.sh:
```diff
@@ -0,0 +1,231 @@
+# This script does spell checks and can add words to a dictionary.
+# It can be used as a git hook to perform checks, it should be named
+# "pre-push","pre-commit" based on the use case and kept into the ".git/hooks/"
+# and be executable.It can be bypassed by--no-verify option with git command.
```
Space before "It".

jonte replied:
and also before "--no-verify"

jonte reviewed (changes requested):
Very nice! :D
I commented for some small improvements, but in general - awesome!

tobsan commented:
@playpeace I'm fine with fixing the spelling in another PR.

Commit:
Add Spell Check capability

- Spell check performed as part of the build step
- Current typos are added to dictionary to avoid build
fail
- Documentation update

Signed-off-by: Sumit Ahuja <ext-sahuja@luxoft.com>

playpeace commented:
@jonte Pushed the changes as suggested.

tobsan commented on docs/check_spelling.sh:
```diff
@@ -0,0 +1,258 @@
+#!/bin/bash
+
+# This file is part of the Software Factory Blueprint project
+# Copyright (C) Pelagicore AB
```
Sorry for being picky, but please add 2017 as year here.

jonte committed:
check_spelling.sh: Add copyright year
